### ISSUE TITLE
Start shipping our own type stubs

### ISSUE THREAD

**Issue by sethmlarson** [urllib3/urllib3#1897]
Typeshed houses very rudimentary type stubs for urllib3 but they're hidden within `requests.packages.urllib3` so Mypy can't take advantage of them when using urllib3 on its own. We could also probably improve on what is already in typeshed by adding better types than `Any` everywhere.

**Comment by judej** [comment]
Seth, are you looking to add type info in the code or add .pyi files with the py files or contribute a more complete set at typeshed? I'd like to help with this

**Comment by savarin** [comment]
Hi @sethmlarson! I'm at Recurse Center and got introduced to his by @shazow. Could I clarify, as per the title, what you have in mind is to have the `.pyi` files alongside `.py` files? Would the first step simply involve porting the one in [requests](https://github.com/python/typeshed/tree/master/third_party/2and3/requests/packages/urllib3), and then expanding on the `Any`?

I've actually been a little stuck getting the stubs hidden in [requests](https://github.com/python/typeshed/tree/master/third_party/2and3/requests/packages/urllib3) to work with `mypy`, even after importing `requests`.

For additional context, my previous role involved making py2 code compatible with py3 and adding type comments. The thought was there's less friction not having to review separate `.pyi` files when refactoring.

**Comment by sethmlarson** [comment]
@savarin Glad you're interested in helping us here.  I was thinking starting with the type stubs that typeshed has (pyi) and making the `Any` more specific.

**Comment by judej** [comment]
@sethmlarson, would you prefer that we modify the stubs in typeshed or would you rather have the .pyi files in this repo checked in alongside the .py files? OR, would you rather add the type info into the .py files?

**Comment by sethmlarson** [comment]
I'd like the pyi files shipped with urllib3.

**Comment by savarin** [comment]
Cool, sounds like a good starting point. Two questions, stemming from my lack of familiarity with stubs and/or `typeshed`.

1. How do you run a type check with the existing `typeshed` [stubs](https://github.com/python/typeshed/tree/master/third_party/2and3/requests/packages/urllib3) of `urllib3` since it's in `requests`? Detailed run below, importing `requests` doesn't seem to help.

```
$ cat run.py

import requests
import urllib3

def pool():
    return urllib3.PoolManager()

$ mypy run.py
run.py:3: error: Skipping analyzing 'urllib3': found module but no type hints or library stubs
run.py:3: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
Found 1 error in 1 file (checked 1 source file)
```

2. I presume all the changes would be in the `typeshed` [repo](https://github.com/python/typeshed/tree/master/third_party/2and3/requests/packages/urllib3) alone. What's the anticipated flow to check these changes are correct (since `typeshed` normally comes with `mypy`), i.e. clone `typeshed`, move to `MYPYPATH`, run `mypy` on `urrlib3` tests (since stubs won't check the code [itself](https://github.com/python/mypy/issues/5028))?

**Comment by pquentin** [comment]
+1 for shipping our own type stubs. `.pyi` files aren't ideal, but types as comments aren't either. Hopefully in 2021 we'll be able to drop Python 2 support in urllib3 2.0.

I have one question though: does anyone understand how providing types for `requests.packages.urllib3` in typeshed achieves anything?

urllib3 used to be vendored in requests, but it's no longer the case: requests now directly imports urllib3, without going through [requests.package.urllib3](https://github.com/psf/requests/blob/master/requests/packages.py) which only exists for backwards-compatibility: it allows others to continue importing urllib3 through `requests.package`.

**Comment by sethmlarson** [comment]
@pquentin I'm guessing it's for backwards compatibility :)

@savarin All changes would be made in the urllib3 repo, not typeshed. I'd recommend the following approach:

- Add the type stubs from typeshed to urllib3 and add necessary changes to ensure pyi files are added to dists

- File by file update all `Any` usages with specific types

- :tada: Profit!

**Comment by savarin** [comment]
> urllib3 used to be vendored in requests, but it's no longer the case

@pquentin Ah this makes more sense now, nullifying questions I had above.

@sethmlarson This is helpful, thank you! After the port, what would the check for correctness be? As above, my understanding is stubs can't run checks on the code so I presume it'll just be on `urllib3/test`.

**Comment by sethmlarson** [comment]
Mypy can check that types are consistent within themselves.

Maybe the approach I took with Eland could be useful for us? https://github.com/elastic/eland/blob/master/noxfile.py#L75

**Comment by savarin** [comment]
Quick follow up - I'm touching a draft first PR, which introduces the mypy check (ignoring inits) and ports the stub files for `packages/ssl_match_hostname`. Full dependency tree [here](https://svgshare.com/i/NtH.svg).

**Comment by savarin** [comment]
@pquentin No worries, just wanted to check it's something you're seeing as well.

I've tested this out a bit more, as discussed commenting out https://github.com/urllib3/urllib3/commit/33c4c13e4d3a5ce77b449b825d8f7214df2e2806 doesn't trigger any mypy errors with `nox -s lint`. I got it to trigger by importing creating a file that imports from `_implementation.py` (compare test passing in https://github.com/urllib3/urllib3/commit/41f39577364206a79c104ef2f320655b38031720 vs failing in https://github.com/urllib3/urllib3/commit/207a5698297dadfbce19b3d057f27a0373b47e72). There's a discussion on a similar issue [here](https://stackoverflow.com/questions/37537194/using-mypy-local-stubs).

@sethmlarson Let me know if you have any thoughts. I'll continue adding stubs and testing it on toy imports. It's a little unsettling that the existing mypy tests aren't raising errors (without the toy import).
